src/Eloquent/Builder.php

```diff
@@ -183,6 +186,41 @@ public function raw($value = null)
         return $results;
     }
 
+    /**
+     * Attempt to create the record if it does not exist with the matching attributes.
+     * If the record exists, it will be returned.
+     *
+     * @param  array $attributes The attributes to check for duplicate records
+     * @param  array $values     The attributes to insert if no matching record is found
+     */
+    public function createOrFirst(array $attributes = [], array $values = []): Model
+    {
+        // Apply casting and default values to the attributes
+        $instance = $this->newModelInstance($values + $attributes);
+        $values = $instance->getAttributes();
+        $attributes = array_intersect_key($attributes, $values);
+
+        return $this->raw(function (Collection $collection) use ($attributes, $values) {
+            $listener = new FindAndModifyCommandSubscriber();
+            $collection->getManager()->addSubscriber($listener);
+
+            try {
+                $document = $collection->findOneAndUpdate(
+                    $attributes,
+                    ['$setOnInsert' => $values],
+                    ['upsert' => true, 'new' => true, 'typeMap' => ['root' => 'array', 'document' => 'array']],
```

Comment:
`new` is an option on the internal FindAndModify operation. Since this is using FindOneAndUpdate, the correct option to use would be `returnDocument` with a value of `FindOneAndUpdate::RETURN_DOCUMENT_AFTER`.